target/consensource-cli
# Assertion Transfer

## Proposed change/fix

This implements the `csrc assertion transfer` command for the cli but the instructions include testing of the [processor pr](https://github.com/target/consensource-processor/pull/17), the [database pr](https://github.com/target/consensource-database/pull/12), and the [sds pr](https://github.com/target/consensource-sds/pull/12).

## Types of changes

What types of changes does this pull request introduce to ConsenSource? _Put an `x` in the boxes that apply_


## How to run/test

1. Clone down the [consensource-compose](https://github.com/target/consensource-compose) repo:
```bash
 git clone --recurse-submodules https://github.com/target/consensource-compose.git
```
2. Make sure each submodule is up-to-date with the latest code changes
```sh
 git submodule update --remote $sds
```
3. `cd sds` and checkout the `tmcdonald-assertion-delete`
4. ~`cd processor` and checkout `pa3ng_add_transfer_assertion `~
5. `cd cli` and checkout `tmcdonald-transfer`
6. Build local images
```sh
 ./docker-helper.sh --build sds cli processor api
```
7. Run `docker-compose -f docker-compose.run-local.yaml up`
8. Exec into the CLI container
9. Create an agent
```bash
 csrc agent create cliagent --url http://api:9009
```
10. Create an ingestion organization
```bash
 csrc organization create assertororg 4 assertercontactname 7632918628 ENG --city minneapolis --country usa --street_address '821 thor street' --url http://api:9009
```
11. Create a factory assertion
```
 csrc assertion factory create {ingestion_org_id_here} testfactoryassertion1 yuki 6123982765 JPN --city osaka --country japan --street_address 'fukuoka south st' --state_province kanto --postal_code 55414 --url http://api:9009
```
12. Create a standard assertion
```
csrc assertion standard create {ingestion_org_id_here} cool_standard "1.0" "desc" "link" 1 --url http://api:9009
```
13. Create a certifcate assertion for the factory to the standard:
```
csrc assertion certificate create {ingestion_org_id_here} {factory_id} 2 3 {standard_id} --url http://api:9009
```

Take note of the returned `assertion id`s for the factory, standard, and certificate.

### Creating 3 additional agents and transferring

While still `exec`ed into the cli image:
```
mkdir -p  /home/agent1/.sawtooth/keys/
mkdir -p  /home/agent2/.sawtooth/keys/
mkdir -p  /home/agent3/.sawtooth/keys/
sawtooth keygen --key-dir=/home/agent1/.sawtooth/keys/
sawtooth keygen --key-dir=/home/agent2/.sawtooth/keys/
sawtooth keygen --key-dir=/home/agent3/.sawtooth/keys/
mv /home/agent1/.sawtooth/keys/root.pub /home/agent1/.sawtooth/keys/agent1.pub
mv /home/agent2/.sawtooth/keys/root.pub /home/agent2/.sawtooth/keys/agent2.pub
mv /home/agent3/.sawtooth/keys/root.pub /home/agent3/.sawtooth/keys/agent3.pub
mv /home/agent1/.sawtooth/keys/root.priv /home/agent1/.sawtooth/keys/agent1.priv
mv /home/agent2/.sawtooth/keys/root.priv /home/agent2/.sawtooth/keys/agent2.priv
mv /home/agent3/.sawtooth/keys/root.priv /home/agent3/.sawtooth/keys/agent3.priv
chmod -R 777 /home/agent1/
chmod -R 777 /home/agent2/
chmod -R 777 /home/agent3/
useradd agent1
useradd agent2
useradd agent3
```
This generates sawtooth keys and user accounts for 3 agents we can use to do transfers.
**Note**: When switching users with the `su` command first exit and verify you are `root`.

### Factory Transfer
1. `su agent1`
2. `cd /cli/target/debug`
3. `./csrc agent create agent1 --url http://api:9009`
4. `./csrc assertion transfer {factory_assertion_id} --url http://api:9009`

After completing the above, ensure the final command works. Then, check the database tables (or api endpoint) to ensure the new owner of the asserted factory is the `other` agent, and that the assertion is no longer returned by the api.

### Standard Transfer

1. `su agent2`
2. `cd /cli/target/debug`
3. `./csrc agent create agent2 --url http://api:9009`
4. `./csrc organization create standard_org 2 agent2 612 en --url http://api:9009`
5. `./csrc assertion transfer {standard_assertion_id} --url http://api:9009`

Verify with db/api the transfer was successful.


#### Certificate Transfer

1. `su agent3`
2. `cd /cli/target/debug`
3. `./csrc agent create agent3 --url http://api:9009`
4. `./csrc organization create auditor 1 agent3 612 en --url http://api:9009`
5. `./csrc assertion transfer {cert_assertion_id} --url http://api:9009`

Verify with db/api the transfer was successful.
